[vcpkg] Download vcpkg.exe rather than building it in bootstrap on Windows.

BillyONeal opened the pull request:
This reduces bootstrap cost for Windows customers, resolving the issue initially submitted as #12502 .

The `toolsrc` tree was extracted to https://github.com/microsoft/vcpkg-tool. `bootstrap.sh` was changed to download the right source tarball, extract, and build it. This was chosen over the previous attempt, a submodule, over concerns of accidentally destroying people's local modifications.

We already have macos signed binaries ready in our infrastructure but the work to edit `bootstrap.sh` to consume them is not yet here.

Next steps:

* <s>Replay any changes to toolsrc in this repo in the tool repo landing after https://github.com/microsoft/vcpkg-tool/commit/913beb74358a66d5de4a7f1eb92b59d16540a17e</s>
* <s>Move / copy infrastructure from this main repo like C++ formatting rules that we want to enforce in the "tool" repo.</s>
* MacOS (in a future PR)
* Investigate doing the same for Linux, e.g. by statically linking everything but glibc and ripping out std::filesystem use. (In future PR(s))

Depends on:

* <s>https://github.com/microsoft/vcpkg/pull/15470</s>
* <s>https://github.com/microsoft/vcpkg/pull/15472</s>
* <s>https://github.com/microsoft/vcpkg/pull/15516</s>
* <s>https://github.com/microsoft/vcpkg/pull/15638</s>
* <s>https://github.com/microsoft/vcpkg/pull/16012</s>

Clean build of vcpkg-tool repo example: https://dev.azure.com/vcpkg/public/_build/results?buildId=48871&view=results

ras0219 commented on scripts/Generate-SpdxLicenseList.ps1:
```diff
@@ -9,10 +9,10 @@ Param(
 
     [Parameter()]
-    [string]$LicensesOutFile = "$PSScriptRoot/../toolsrc/src/vcpkg/spdx-licenses.inc",
+    [string]$LicensesOutFile = "$PSScriptRoot/../vcpkg-tool/src/vcpkg/spdx-licenses.inc",
```
This feels like it should perhaps be extracted into the separate repo.

ras0219 commented on scripts/azure-pipelines/Format-CxxCode.ps1:
```diff
@@ -30,7 +30,7 @@ if ([String]::IsNullOrEmpty($clangFormat) -or -not (Test-Path $clangFormat))
     throw
```
It seems like this file should move to the new repo

ras0219 commented on scripts/azure-pipelines/azure-pipelines.yml:
```diff
@@ -20,12 +20,16 @@ stages:
+    #   inputs:
+    #     filePath: 'scripts/azure-pipelines/Format-CxxCode.ps1'
+    #     arguments: '-Root .'
+    #     pwsh: true
```
Commented code

ras0219 commented on scripts/bootstrap.ps1:
```diff
@@ -57,362 +26,15 @@ while (!($vcpkgRootDir -eq "") -and !(Test-Path "$vcpkgRootDir\.vcpkg-root"))
-if ($ec -ne 0)
-{
-    Write-Error "Building vcpkg.exe failed. Please ensure you have installed Visual Studio with the Desktop C++ workload and the Windows SDK for Desktop C++."
+    Write-Error "Downloadiong vcpkg.exe failed. Please check your internet connection, or consider downloading a recent vcpkg.exe from https://github.com/microsoft/vcpkg-tool with a browser."
```
```suggestion
    Write-Error "Downloading vcpkg.exe failed. Please check your internet connection or download it directly from https://github.com/microsoft/vcpkg-tool/releases/download/2021.01.04-a78fe22da82ca4fb9a92b875d5fd34d4016d5bf0/vcpkg.exe and place it in the root of this repository."
```
(Also it would be nice to then lift out the shared url to `$url` or similar)

ras0219 commented on scripts/bootstrap.sh:
```diff
@@ -240,6 +240,8 @@ selectCXX()
 UNAME="$(uname)"
 ARCH="$(uname -m)"
 
+git submodule update --init vcpkg-tool
```
We should double-check that this always does the right thing from any state (outstanding edits, different checked out version, etc).

JackBoosY commented:
Closes #14995.

StarGate-One commented:
@BillyONeal  - Thanks a lot for the hard work so far.👍👍👍

So to confirm/verify:
1. The `vcpkg-tool` repository replaces the `[vcpkg-root]/toolsrc directory`?
2. Any changes to the `vcpkg` binary will need a PR will be submitted to the `vcpkg-tool` repository or the issue reported on `vcpkg` repository and PR on `vcpkg-tool` repository linked to the vcpkg repository issue or ...?
3. Will there be scripts to build the vcpkg-tool, a bootstrap-vcpkg-tool or something to that effect?

Again thanks a lot for all the work you do!👏👏👏
Have a wonderful day and super week!😎😎😎

BillyONeal commented:
>The vcpkg-tool repository replaces the [vcpkg-root]/toolsrc directory?

Correct. I prepared this using `git filter-branch`.

>Any changes to the vcpkg binary will need a PR will be submitted to the vcpkg-tool repository

Correct. Issue tracking remains here.

>Will there be scripts to build the vcpkg-tool, a bootstrap-vcpkg-tool or something to that effect?

In that repo, it's just a cmake project like any other, so there are no specific scripts. Here, bootstrap on the POSIX platforms still needs to be fixed to download and build the sources.

BillyONeal committed:
[vcpkg] Download vcpkg.exe rather than building it in bootstrap on Windows.

This reduces bootstrap cost for Windows customers, resolving the issue initially submitted as #12502 .

The toolsrc tree was extracted to https://github.com/microsoft/vcpkg-tool, and created as a submodule "vcpkg-tool" here. The name is explicitly chosen to be different from toolsrc because git doesn't like an ordinary directory becoming a submodule when changing branches or vice versa. Given that the name had to change, I set the name to the same name as the actual github repo.

Clean build of vcpkg-tool repo example: https://dev.azure.com/vcpkg/public/_build/results?buildId=48871&view=results

BillyONeal committed:
Remove nonexistent e2e tests on macos.

BillyONeal committed:
Merge remote-tracking branch 'origin/master' into downloads_on_windows

# Conflicts:
#	scripts/azure-pipelines/end-to-end-tests-dir/registries.ps1
#	toolsrc/cmake/utilities.cmake
#	toolsrc/include/vcpkg/base/expected.h
#	toolsrc/include/vcpkg/vcpkgcmdarguments.h
#	toolsrc/src/vcpkg-test/binarycaching.cpp
#	toolsrc/src/vcpkg-test/json.cpp
#	toolsrc/src/vcpkg-test/manifests.cpp
#	toolsrc/src/vcpkg.cpp
#	toolsrc/src/vcpkg/base/json.cpp
#	toolsrc/src/vcpkg/base/parse.cpp
#	toolsrc/src/vcpkg/build.cpp
#	toolsrc/src/vcpkg/dependencies.cpp
#	toolsrc/src/vcpkg/registries.cpp
#	toolsrc/src/vcpkg/vcpkgcmdarguments.cpp

BillyONeal committed:
Update vcpkg-tool version.

ras0219-msft commented on docs/tool-maintainers/benchmarking.md:
```diff
@@ -3,7 +3,7 @@
 large changes to how something works. If you are attempting to make something
 faster, and you end up slowing it down, you'll never know if you don't
-benchmark! We have benchmarks in the `toolsrc/src/vcpkg-test` directory, just
+benchmark! We have benchmarks in the `vcpkg-tool/src/vcpkg-test` directory, just
```
These seem out of date without a submodule.

BillyONeal committed:
Transfer .\docs\maintainers

BillyONeal committed:
Bump vcpkg-tool version.

BillyONeal committed:
Merge remote-tracking branch 'origin/master' into downloads_on_windows

strega-nil commented on scripts/bootstrap.sh:
```diff
@@ -276,11 +276,29 @@ else
 # Do the build
-buildDir="$vcpkgRootDir/toolsrc/build.rel"
-rm -rf "$buildDir"
+vcpkgToolReleaseTag="2021-01-13-768d8f95c9e752603d2c5901c7a7c7fbdb08af35"
+vcpkgToolReleaseSha="99c9949637f83bf361ee23557edc889e2865c2105c45306c39c40855a3e1440e2f6fb5ec59e95176fca61eff33929462d23c7b49feb1975f24adb8ca443a98a6"
```
Hrm, why are these different? I would expect that the release tag should have the same end part as the sha.

strega-nil commented on scripts/bootstrap.sh:
```diff
@@ -276,11 +276,29 @@ else
+srcDir="$srcBaseDir/vcpkg-tool-$vcpkgToolReleaseTag"
+
+if [ -e "$tarballPath" ]; then
+    vcpkgCheckEqualFileHash "$vcpkgToolUrl" "$tarballPath" "$vcpkgToolReleaseSha"
```
hmm, we should probably check the file hash, and if they're not equal, delete it and download again, as opposed to just failing.

BillyONeal commented:
Considering the status quo for all the other downloads in this script (or elsewhere in the tool -- that's why we had to add `FILE_DISAMBIGUATOR` to places) doesn't do that would you be OK just turning that into an issue?

strega-nil commented:
@BillyONeal yep, I'm good with that.
